### ISSUE TITLE
i/o timeout host.docker.internal when two static DNS server IP addresses added to host

### ISSUE THREAD

**Issue by johnthagen** [moby/moby#48266]
### Description

We have been tracking down a very obscure slow down to our web application that runs under Docker Community Edition (CE) and Docker Compose. We finally narrowed the issue down to hosts that have two (or more?) static DNS server IP addresses set. This causes DNS requests to be sent from `dockerd` for `host.docker.internal`. On some machines, we noticed multiple 4 second pauses in `dockerd` due to this that would greatly slow down the response time of a web back running within one of the Docker Compose containers.

### Reproduce

Ubuntu 22.04.4 machine or VM.

1. Install Docker Engine manually using the steps here: https://docs.docker.com/engine/install/ubuntu/#install-from-a-package
2. Set two static DNS server IP addresses through the Network Settings. In this example we use Google's 8.8.8.8 and 8.8.4.4, but this problem has been observed with other IP addresses.

<img width="975" alt="Screenshot 2024-07-29 at 3 31 05 PM" src="https://github.com/user-attachments/assets/05ff532d-ddb0-4c88-b5f3-ac479cfc8320">

3. Restart the system
4. Search the `dockerd` syslogs (in our case we have a Docker Compose project with `restart: unless-stopped`)

```
$ sudo grep dockerd /var/log/syslog
...
Jul 29 15:23:57 ub2204 dockerd[1229]: time="2024-07-29T15:23:57.540411785-04:00" level=error msg="[resolver] failed to query external DNS server" client-addr="udp:127.0.0.1:58726" dns-server="udp:127.0.0.53:53" error="read udp 127.0.0.1:58726->127.0.0.53:53: i/o timeout" question=";host.docker.internal.\tIN\t A" spanID=ad07ae9a38ccecfb traceID=edfbb74c0ab95a42ffcb6721f96aad66
```

On some systems these will block for 4 seconds before timing out, blocking the container from responding to web requests. Nothing in our application refers to `host.docker.internal`.

### Expected behavior

Docker Community Edition should not reference `host.docker.internal`. According to the docs, this is only a (paid) Docker Desktop feature. Why is Docker Community Edition referencing it?

- https://docs.docker.com/desktop/networking/#i-want-to-connect-from-a-container-to-a-service-on-the-host

This stackoverflow answer (https://stackoverflow.com/a/43541681) says:

> Use your internal IP address or connect to the special DNS name host.docker.internal which will resolve to the internal IP address used by the host.

> This is for development purpose and does not work in a production environment outside of Docker Desktop.

Why is `host.docker.internal` being referenced by Docker CE? Docker Desktop has not been installed on these systems.

### docker version

```bash
$ docker version
Client: Docker Engine - Community
 Version:           26.1.4
 API version:       1.45
 Go version:        go1.21.11
 Git commit:        5650f9b
 Built:             Wed Jun  5 11:28:57 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          26.1.4
  API version:      1.45 (minimum version 1.24)
  Go version:       go1.21.11
  Git commit:       de5c9cf
  Built:            Wed Jun  5 11:28:57 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.33
  GitCommit:        d2d58213f83a351ca8f528a95fbd145f5654e957
 runc:
  Version:          1.1.12
  GitCommit:        v1.1.12-0-g51d5e94
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

### docker info

```bash
$ docker info
Client: Docker Engine - Community
 Version:    26.1.4
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  v2.27.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 26.1.4
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: d2d58213f83a351ca8f528a95fbd145f5654e957
 runc version: v1.1.12-0-g51d5e94
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-41-generic
 Operating System: Ubuntu 22.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 15.57GiB
 Name: ub2204
 ID: 7ecbea20-6ef6-4a54-83e6-a66e62ca30cb
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by johnthagen** [comment]
I could not reproduce this on the following newer version. But since our current application pins the version of Docker in the previous post, it would be helpful to hear some commentary about if this was something intentionally fixed?

```
docker version
Client: Docker Engine - Community
 Version:           27.0.3
 API version:       1.46
 Go version:        go1.21.11
 Git commit:        7d4bcd8
 Built:             Sat Jun 29 00:02:33 2024
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          27.0.3
  API version:      1.46 (minimum version 1.24)
  Go version:       go1.21.11
  Git commit:       662f78c
  Built:            Sat Jun 29 00:02:33 2024
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.7.19
  GitCommit:        2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc:
  Version:          1.7.19
  GitCommit:        v1.1.13-0-g58aa920
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

```
docker info
Client: Docker Engine - Community
 Version:    27.0.3
 Context:    default
 Debug Mode: false
 Plugins:
  compose: Docker Compose (Docker Inc.)
    Version:  v2.29.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 4
  Running: 4
  Paused: 0
  Stopped: 0
 Images: 4
 Server Version: 27.0.3
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local splunk syslog
 Swarm: inactive
 Runtimes: runc io.containerd.runc.v2
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2bf793ef6dc9a18e00cb12efb64355c2c9d5eb41
 runc version: v1.1.13-0-g58aa920
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.5.0-41-generic
 Operating System: Ubuntu 22.04.4 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 16
 Total Memory: 15.57GiB
 Name: ub2204
 ID: dbb076f5-7613-413f-a948-8655e3075ce1
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
The daemon itself doesn't have code to perform these queries, but the daemon does have an embedded DNS resolver that is used for resolving hostnames of containers in the same docker network, and acts as forwarder for containers; the logs you found are from that embedded DNS resolver;

```
Jul 29 15:23:57 ub2204 dockerd[1229]: time="2024-07-29T15:23:57.540411785-04:00" level=error msg="[resolver] failed to query external DNS server" client-addr="udp:127.0.0.1:58726" dns-server="udp:127.0.0.53:53" error="read udp 127.0.0.1:58726->127.0.0.53:53: i/o timeout" question=";host.docker.internal.\tIN\t A" spanID=ad07ae9a38ccecfb traceID=edfbb74c0ab95a42ffcb6721f96aad66
```

Those logs indicate that something in a container is trying to resolve `host.docker.internal`, and the docker daemon forwarding this request to the DNS server configured on your host (in this case, likely systemd-resolvd (127.0.0.53)).


> Why is `host.docker.internal` being referenced by Docker CE? Docker Desktop has not been installed on these systems.

It's possible you're using an image that has this hostname configured as a default somewhere; you mention you are running a compose project ("in our case we have a Docker Compose project with `restart: unless-stopped`"); can you share this project? What images are used?

**Comment by thaJeztah** [comment]
> I could not reproduce this on the following newer version. But since our current application pins the version of Docker in the previous post, it would be helpful to hear some commentary about if this was something intentionally fixed?

There have been some fixes around resolving DNS from containers running with a network with `internal` set,  and some improvements around DNS resolutions; it's possible that these requests no longer produce the error.

Also note that `host.docker.internal` can be configured as a custom "extra-host" in containers, in which case the hostname is added to the container's `/etc/hosts`, but this should be visible in your compose-file, or by inspecting containers that are running.

**Comment by johnthagen** [comment]
@thaJeztah Thank you for the reply and useful details.

> can you share this project? What images are used?

Unfortunately the project is private, but I will share some minimized details.

```yaml
services:
  backend:
    build: <Custom Dockerfile `FROM python:3.11-slim-bookworm` Django application>
    command: gunicorn ...
    networks:
      - frontend
      - database
      - custom
    expose:
      - 8000
    depends_on:
      - postgres
      - custom
      - frontend
    restart: unless-stopped
  postgres:
    image: postgres:13-alpine
    networks:
      - database
    expose:
      - 5432
    restart: unless-stopped
  custom:
    build: <Custom Dockerfile `FROM python:3.11-slim-bookworm` FastAPI application>
    command: uvicorn ...
    networks:
      - custom
    expose:
      - 8008
    restart: unless-stopped
  frontend:
    build: <Custom Dockerfile `FROM nginx:1.25-alpine` Serves static files and reverse proxy to backend>
    networks:
      - frontend
    ports:
      - "443:443"
      - "80:80"
    restart: unless-stopped

networks:
  frontend:
  database:
  custom:
  
 # (volumes omitted)
```

None of our application code references `host.docker.internal`. Perhaps somehow the `postgres:` or `nginx:` images do?

**Comment by johnthagen** [comment]
@thaJeztah Well, would you look at that. I found this in one of our Python dependencies:

https://github.com/jazzband/django-debug-toolbar/blob/c5cdd70d8f143a1e6c526896fcba808a75a6dcce/debug_toolbar/middleware.py#L36


```py
    # Test: Docker
    try:
        # This is a hack for docker installations. It attempts to look
        # up the IP address of the docker host.
        # This is not guaranteed to work.
        docker_ip = (
            # Convert the last segment of the IP address to be .1
            ".".join(socket.gethostbyname("host.docker.internal").rsplit(".")[:-1])
            + ".1"
        )
        if request.META.get("REMOTE_ADDR") == docker_ip:
            return True
```

**Comment by johnthagen** [comment]
Related issue

- https://github.com/jazzband/django-debug-toolbar/issues/1927

**Comment by akerouanton** [comment]
Hi @johnthagen,

You edited your last comment while I was writing this... but let me post it.

The error log is accurate. Let me break it down:

```
msg="[resolver] failed to query external DNS server"
client-addr="udp:127.0.0.1:58726"
dns-server="udp:127.0.0.53:53"
error="read udp 127.0.0.1:58726->127.0.0.53:53: i/o timeout" question=";host.docker.internal.\tIN\t A"
```

Here, the embedded DNS server tries to contact a DNS server located at `127.0.0.53` on port 53 but that server doesn't answer. Since it's UDP, it might be unresponsive or not running at all.

When the Engine created your containers, it took a look at your host `/etc/resolv.conf` and extracted the nameserver field to configure the embedded DNS server. That's where this `127.0.0.53` is coming from.

If the daemon can't talk to that DNS server, it means there's probably something broken on your computer / server.

I'll close this ticket as this bug isn't on our end but feel free to continue the discussion.

**Comment by johnthagen** [comment]
@akerouanton  I am happy to close this issue. It was a client-side issue with code within the container.

It does manifest differently on different Docker versions, but I think that is unrelated.

Thank you @thaJeztah for the information that helped me track down the root cause of the issue.

**Comment by akerouanton** [comment]
Woops, I posted an earlier comment saying this issue wasn't on our end but I now realize that `host.docker.internal` should not be forwarded to upstream DNS servers 🤔 Let me reopen it.

**Comment by thaJeztah** [comment]
@akerouanton we'd probably need to double check how the desktop team implemented it; it's possible they resolve it in docker desktop's DNS forwarder? (which would be "external"?)

**Comment by akerouanton** [comment]
I was pretty sure we were automatically adding the extra host `host.docker.internal`, but that's not the case. Engine users need to use `--extra-host=host.docker.internal:host-gateway` to get it.

The current behavior is to forward DNS queries upstream if no matching records can be found locally. So, what you observed is expected.

But I think we should do better: the Engine's split-brain / embedded DNS server should be the authority for the `docker.internal.` zone and if no extra-host is passed, it should return a negative answer for `host.docker.internal.` (or any other record it doesn't know). That would make sure no `docker.internal.` queries are forwarded uptsream.

Fortunately, [ICANN proceeding 'Proposed Top-Level Domain String for Private Use'](https://www.icann.org/en/public-comment/proceeding/proposed-top-level-domain-string-for-private-use-24-01-2024) passed and settled on reserving `.internal.`. However, this isn't yet part of [IANA 'Special-Use Domain Names' registry](https://www.iana.org/assignments/special-use-domain-names/special-use-domain-names.xhtml), and it's unclear when it'll be. Once it's there, we should really look at making the embedded DNS resolver the authority for `.docker.internal.`.

My yesterday's answer was accurate, let me re-post it as is. Note that, while your issue has been fixed by other means, you might have an underlying DNS issue on your host, or your local DNS forwarder might block requests for the `.internal.` zone.

Also, I ran v26.1.4 with breakpoints set, and I can confirm the DNS query is forwarded upstream too. I believe the discripancy you witnessed between both versions was due to something else, probably your local DNS forwarder, but not the Engine.

> The error log is accurate. Let me break it down:
> 
> ```
> msg="[resolver] failed to query external DNS server"
> client-addr="udp:127.0.0.1:58726"
> dns-server="udp:127.0.0.53:53"
> error="read udp 127.0.0.1:58726->127.0.0.53:53: i/o timeout" question=";host.docker.internal.\tIN\t A"
> ```
>
> Here, the embedded DNS server tries to contact a DNS server located at 127.0.0.53 on port 53 but that server doesn't answer. Since it's UDP, it might be unresponsive or not running at all.
> 
> When the Engine created your containers, it took a look at your host /etc/resolv.conf and extracted the nameserver field to configure the embedded DNS server. That's where this 127.0.0.53 is coming from.
> 
> If the daemon can't talk to that DNS server, it means there's probably something broken on your computer / server.
> 
> I'll close this ticket as this bug isn't on our end but feel free to continue the discussion.
